src/livestatus.c

```diff
@@ -0,0 +1,463 @@
+    return -1;
+  }
+
+  memset(&sun, 0x0, sizeof(sun));
+  sun.sun_family = AF_UNIX;
+  strcpy(sun.sun_path, sockfile);
+
+again:
+  rc = connect(sfd, (struct sockaddr *)&sun, sizeof(sun));
+  if (rc < 0) {
+    if (errno != EINTR) {
+      ERROR("livestatus plugin: connect to socket file %s: %s", sockfile,
+            STRERRNO);
+      return -1;
+    }
+    goto again;
+  }
+
+  *sockfd = sfd;
+
+  return 0;
+} /* unix_connect */
+
+static int ls_parse(const char *lresponse, livestatus_status_t *lstatus) {
+  int rc = -1;
+  int i = 0;
+  char **fields = NULL;
+  char *pchar = NULL;
+
+  fields = malloc(sizeof(char **) * LIVESTATUS_EXPECTED_FIELDS_RESP_NB);
+  if (fields == NULL) {
+    ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+    return -1;
+  }
+
+  for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+    fields[j] = malloc(sizeof(char *) * 32);
+    if (fields[j] == NULL) {
+      ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+      rc = -1;
+      goto free_all_fields;
+    }
+
+    memset(fields[j], 0x0, 32);
+  }
+
+  pchar = (char *)lresponse;
+  char *wstart = (char *)lresponse;
+  int exit = 0;
+  while (!exit) {
+    if (*pchar == '\0' || *pchar == '\n') {
+      exit = 1;
+    } else {
+      if (*pchar != ';') {
+        pchar++;
+        continue;
+      }
+    }
+
+    strncpy(fields[i], wstart, pchar - wstart);
+    wstart = pchar + 1;
+    i++;
+    pchar++;
+
+    if (!exit) {
+      if (i >= LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+        ERROR("livestatus plugin: too many fields in livestatus output");
+        rc = -2;
+        goto free_all_fields;
+      }
+    }
+  }
+
+  if (i < LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+    ERROR("livestatus plugin: not enough fields in livestatus output");
+    rc = -3;
+    goto free_all_fields;
+  }
+
+  *lstatus = c_to_livestatus_status((const char **)fields);
+
+  rc = 0;
+
+free_all_fields:
+  for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+    if (fields[j] != NULL) {
+      free(fields[j]);
+      fields[j] = NULL;
+    }
+  }
+
+  free(fields);
+  fields = NULL;
+
+  return rc;
+} /* int ls_parse */
+
+static int ls_read_parse(const int sockfd, livestatus_status_t *lstatus) {
+  ssize_t bread = -1;
+  char buffer[IO_BUFFER_SIZE];
+
+again:
+  memset(&buffer, 0x0, sizeof(buffer));
+
+  bread = read((int)sockfd, &buffer, IO_BUFFER_SIZE - 1);
+  if (bread < 0) {
+    if (errno != EINTR) {
+      ERROR("livestatus plugin: reading from socket: %s", STRERRNO);
+      return -1;
+    }
+    goto again;
+  }
+
+  buffer[IO_BUFFER_SIZE - 1] = '\0';
+
+  return ls_parse((const char *)&buffer, lstatus);
+} /* int ls_read_parse */
+
+static int ls_send_request(const int sockfd) {
+  int written = -1;
+  char request[512];
+
+  memset(&request, 0x0, sizeof(request));
+
+  snprintf(request, sizeof(request) - 1, "GET status\nColumns: %s\n\n",
+           LIVESTATUS_QUERY_COLUMNS);
+
+again:
```

Comment:
goto is sometimes used for error handling, but here it is more typical use for a while loop.

Reply:
I've migrated to a `do{...}while(errno == EINTR)` pattern. Is this okay for you ?